File: dictator/mcp_client_tool.py
#!/usr/bin/env python3
import sys
import json
import asyncio
from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client

async def main():
    if len(sys.argv) < 3:
        print("Usage: mcp_client_tool.py <tool_name> [tool_args_json]")
        sys.exit(1)

    tool_name = sys.argv[1]
    tool_args = json.loads(sys.argv[2]) if len(sys.argv) > 2 else {}

    # This client is designed to be called BY a worker (legion) 
    # to talk to the Dictator (MCP Server) via stdio.
    # However, in ROME, the Dictator is the one calling the worker.
    # This tool is for cross-tool communication if needed.
    
    server_params = StdioServerParameters(
        command="python3",
        args=["/home/paul-kane/projects/rome-core/dictator/dictator.py"],
        env=None
    )

    async with stdio_client(server_params) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()
            result = await session.call_tool(tool_name, arguments=tool_args)
            
            for item in result.content:
                if item.type == "text":
                    print(item.text)

if __name__ == "__main__":
    asyncio.run(main())

File: dictator/mcp_client_tool.mjs
#!/usr/bin/env node

import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import { McpClient } from "@modelcontextprotocol/sdk/client/mcp.js";
import { z } from "zod";

async function main() {
  const argsSchema = z.object({
    tool_name: z.string(),
    tool_args: z.string().optional().default("{}").transform((str) => JSON.parse(str)),
  });

  const parsedArgs = argsSchema.parse({
    tool_name: process.argv[2],
    tool_args: process.argv[3],
  });

  const transport = new StdioClientTransport();
  const client = new McpClient(transport);
  await client.connect();

  const result = await client.toolCall(parsedArgs.tool_name, parsedArgs.tool_args);

  // MCP server responds with an array of content objects.
  // We need to extract the text content and print it for legion_wrapper to capture.
  if (result.content && result.content.length > 0) {
    for (const item of result.content) {
      if (item.type === "text") {
        console.log(item.text);
      }
      // Handle other types if necessary, though for now, we expect JSON text
    }
  } else if (result.error) {
    console.error(`MCP Client Error: ${result.error.message}`);
  }

  client.disconnect();
}

main().catch((error) => {
  console.error("MCP Client Wrapper crashed:", error);
  process.exit(1);
});
